web-services/query/src/main/java/datawave/webservice/query/limit/QueryLimiter.java

```diff
@@ -0,0 +1,570 @@
+    }
+
+    /**
+     * Set the hostname provider. This only needs to be done for testing purposes when we want to avoid using {@link InetHostnameProvider}.
+     *
+     * @param provider
+     *            the hostname provider
+     */
+    public void setHostnameProvider(HostnameProvider provider) {
+        this.hostnameProvider = provider;
+    }
+
+    /**
+     * Return the {@link ActiveQueryTracker} instance, initializing it if needed.
+     *
+     * @return the active query tracker
+     */
+    private ActiveQueryTracker getActiveQueryTracker() throws QuorumPeerConfig.ConfigException {
+        if (this.activeQueryTracker == null) {
+            this.activeQueryTracker = new ActiveQueryTracker(zookeeperConfig, 120000L);
+        }
+        return this.activeQueryTracker;
+    }
+
+    /**
+     * Evaluates a {@link ActiveQuerySnapshot} to see if it meets any limits.
+     */
+    private class LimitChecker {
+
+        private final String userDn;
+        private final String system;
+        private final String queryLogic;
+        private ActiveQuerySnapshot snapshot;
+
+        // Whether the user has a custom query logic group limit.
+        private boolean userOverrodeQueryLogicLimit;
+
+        // Whether a limit was met.
+        private boolean metLimit;
+
+        // The message to return.
+        private String message;
+
+        public LimitChecker(String userDn, String system, String queryLogic) {
+            this.userDn = userDn;
+            this.system = system;
+            this.queryLogic = queryLogic;
+        }
+
+        /**
+         * Check the limits against the snapshot.
+         */
+        public void checkLimits() throws Exception {
+            loadSnapshot();
+
+            // Check limits configured for the user.
+            checkUserLimits();
+            if (metLimit) {
+                return;
+            }
+
+            // Check limits configured for the system.
+            checkSystemLimits();
+            if (metLimit) {
+                return;
+            }
+
+            // Check limits configured for the query logic.
+            checkQueryLogicLimits();
+        }
+
+        /**
+         * Load the snapshot of relevant active queries.
+         *
+         * @throws Exception
+         *             if an error occurs while fetching the snapshot.
+         */
+        private void loadSnapshot() throws Exception {
+            // When fetching the snapshot of query activity, we need to include queries for any query logics that may count towards a query logic group limit
+            // applicable for the candidate user, system, and query logic. The applicable query logic groups depend on the default query logic group limits,
+            // along with any custom limits defined for the user or system.
+            Set<String> queryLogics = getQueryLogicsToFetchInSnapshot();
+            this.snapshot = getActiveQueryTracker().getSnapshot(userDn, system, queryLogics);
```

Comment:
Do we need to load everything in the snapshot for the user, system, and query logics?  Perhaps we only need a subset of that depending on what the configured limits are?